File: src/leela/server/network/protocol.py
#!/usr/bin/python
# -*- coding: utf-8; -*-
#
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

from leela.server.data import event

def serialize_event(e):
    return("event %d|%s %d.0 %s;" % (len(e.name()), e.name(), e.unixtimestamp(), repr(e.value())))

def serialize_events(es):
    return("".join(map(serialize_event, es)))

def serialize_proc(cmd):
    return("proc %s;" % cmd)

def serialize_close():
    return("close;")

def parse_event(l):
    if (l.startswith("event ")):
        s  = l.index("|")
        k  = int(l[6:s], 10)
        n  = l[s+1:s+1+k]
        k  = l.index(" ", s+1+k)
        k1 = l.index(" ", k+1)
        t  = float(l[k+1:k1])
        if (l.endswith(";")):
            v = float(l[k1+1:-1])
        else:
            v = float(l[k1+1:])
        return(event.Event(n, v, t))
    else:
        raise(RuntimeError("error: bad format"))

def parse_events(l):
    not_null = lambda s: s!=""
    return(map(parse_event, filter(not_null, l.split(";"))))

File: src/core/DarkMatter/Network/Protocol.hs
-- All Rights Reserved.
--
--    Licensed under the Apache License, Version 2.0 (the "License");
--    you may not use this file except in compliance with the License.
--    You may obtain a copy of the License at
--
--        http://www.apache.org/licenses/LICENSE-2.0
--
--    Unless required by applicable law or agreed to in writing, software
--    distributed under the License is distributed on an "AS IS" BASIS,
--    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
--    See the License for the specific language governing permissions and
--    limitations under the License.

-- | We are using unix sockets + SEQPACKET so there is no need to
-- defined a network protocol. The only thing this does is define the
-- buffer size.
module DarkMatter.Network.Protocol where

import           Data.Bits
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as B8
import           Network.Socket (Socket)
import           Network.Socket.ByteString
import           DarkMatter.Data.Asm.Types (Key)
import           DarkMatter.Data.Asm.Parser
import           DarkMatter.Data.Event
import           DarkMatter.Data.Proc

data Status = Success
            | Failure

unpackShort :: B.ByteString -> Int
unpackShort n = let [a, b] = map fromIntegral (B.unpack n)
                in (a `shiftL` 8) .|. b

packShort :: Int -> B.ByteString
packShort i0 = let a = fromIntegral (i0 `shiftR` 8 .&. 0xFF)
                   b = fromIntegral (i0 .&. 0xFF)
               in B.pack [a, b]

recvFrame :: Socket -> IO B.ByteString
recvFrame s = do { sz <- recv s 2
                 ; if (B.length sz == 2)
                   then recv s (unpackShort sz)
                   else return B.empty
                 }

sendFrame :: Socket -> B.ByteString -> IO ()
sendFrame s msg = let sz = B.length msg
                  in sendAll s (packShort sz `B.append` msg)

sendStatus :: Socket -> Status -> IO ()
sendStatus s Success = sendFrame s (B8.pack "status 0;")
sendStatus s Failure = sendFrame s (B8.pack "status 1;")

eol :: B.ByteString
eol = B8.singleton ';'

databusParser :: Proc B.ByteString [(Key, Event)]
databusParser = Auto $ f B.empty
  where f l r = let (frame, leftover) = B8.breakEnd (== ';') (B.append l r)
                    events            = runAll eventParser frame
                in (events , Auto $ f leftover)

